silkworm/db/snapshots/index/btree_index.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#pragma once
+
+#include <cstdint>
+#include <filesystem>
+#include <memory>
+#include <optional>
+
+#include <silkworm/infra/common/memory_mapped_file.hpp>
+
+#include "../common/encoding/elias_fano.hpp"
+#include "../seg/decompressor.hpp"
+#include "btree.hpp"
+
+namespace silkworm::snapshots::index {
+
+using encoding::EliasFanoList32;
+
+class BTreeIndex {
+  public:
+    static constexpr auto kDefaultFanout{256};
+
+    using DataIndex = BTree::DataIndex;
+    using DataIterator = BTree::DataIterator;
+
+    class Cursor {
+      public:
+        ByteView key() const noexcept { return key_; }
+        ByteView value() const noexcept { return value_; }
+        DataIndex data_index() const noexcept { return data_index_; }
+
+        bool next();
+
+      private:
+        friend class BTreeIndex;
+
+        Cursor(BTreeIndex* index, ByteView key, ByteView value, DataIndex data_index, DataIterator data_it);
+        bool to_next();
+
+        BTreeIndex* index_;
```

Comment:
No, currently it is not because `seek` returns `std::optional<Cursor>` and such change would not compile (i.e. if `Cursor` contains a reference becomes no more copy-aasignable)

Reply:
Chat GPT says that optional doesn't require CopyAssignable: 

> In C++, std::optional<T> does indeed impose certain requirements on T. Specifically, T must be CopyConstructible or MoveConstructible for std::optional<T> to function properly, but T is not required to be CopyAssignable or MoveAssignable in general, unless you need to use specific operations that imply assignment.